executor/src/witgen/jit/symbolic_expression.rs

```diff
@@ -0,0 +1,314 @@
+
+    fn mul(self, rhs: Self) -> Self::Output {
+        if let (SymbolicExpression::Concrete(a), SymbolicExpression::Concrete(b)) = (self, rhs) {
+            SymbolicExpression::Concrete(*a * *b)
+        } else if self.is_known_zero() || rhs.is_known_zero() {
+            SymbolicExpression::Concrete(T::from(0))
+        } else if self.is_known_one() {
+            rhs.clone()
+        } else if rhs.is_known_one() {
+            self.clone()
+        } else if self.is_known_minus_one() {
+            -rhs
+        } else if rhs.is_known_minus_one() {
+            -self
+        } else {
+            SymbolicExpression::BinaryOperation(
+                Box::new(self.clone()),
+                BinaryOperator::Mul,
+                Box::new(rhs.clone()),
+                None,
+            )
+        }
+    }
+}
+
+impl<T: FieldElement, V: Clone> Mul for SymbolicExpression<T, V> {
+    type Output = SymbolicExpression<T, V>;
+    fn mul(self, rhs: Self) -> Self {
+        &self * &rhs
+    }
+}
+
+impl<T: FieldElement, V: Clone> SymbolicExpression<T, V> {
+    /// Field element division. See `integer_div` for integer division.
+    /// If you use this, you must ensure that the divisor is not zero.
+    pub fn field_div(&self, rhs: &Self) -> Self {
+        if let (SymbolicExpression::Concrete(a), SymbolicExpression::Concrete(b)) = (self, rhs) {
+            assert!(b != &T::from(0));
+            SymbolicExpression::Concrete(*a / *b)
+        } else if self.is_known_zero() {
+            SymbolicExpression::Concrete(T::from(0))
+        } else if rhs.is_known_one() {
+            self.clone()
+        } else if rhs.is_known_minus_one() {
+            -self
+        } else {
+            // TODO other simplifications like `-x / -y => x / y`, `-x / concrete => x / -concrete`, etc.
+            SymbolicExpression::BinaryOperation(
+                Box::new(self.clone()),
+                BinaryOperator::Div,
+                Box::new(rhs.clone()),
+                None,
+            )
+        }
+    }
+
+    /// Integer division, i.e. convert field elements to unsigned integer and divide.
+    pub fn integer_div(&self, rhs: &Self) -> Self {
+        if rhs.is_known_one() {
+            self.clone()
+        } else {
+            SymbolicExpression::BinaryOperation(
+                Box::new(self.clone()),
+                BinaryOperator::IntegerDiv,
+                Box::new(rhs.clone()),
+                None,
+            )
+        }
+    }
+}
+
+impl<T: FieldElement, V: Clone> BitAnd for &SymbolicExpression<T, V> {
+    type Output = SymbolicExpression<T, V>;
+
+    fn bitand(self, rhs: Self) -> Self::Output {
+        if let (SymbolicExpression::Concrete(a), SymbolicExpression::Concrete(b)) = (self, rhs) {
+            SymbolicExpression::Concrete(T::from(a.to_integer() & b.to_integer()))
+        } else if self.is_known_zero() || rhs.is_known_zero() {
+            SymbolicExpression::Concrete(T::from(0))
+        } else {
```

Comment:
You mean check if `self` or `rhs` has "all bits set" and in that case, return the other value?

That is tricky because we are usually dealing with prime fields and if we have "all bits set", we are outside of the field.